### PR TITLE
ARROW-1654: [Python] Implement pickling for DataType, Field, Schema

### PR DESCRIPTION
**Pull request by wesm**

### CODE REVIEW

**Comment by wesm** [comment]
This is completed

File: python/pyarrow/tests/test_schema.py [hunk 1]
```diff
@@ -304,3 +306,37 @@ def test_schema_repr_with_dictionaries():
...
+        pa.binary(10),
+        pa.list_(pa.string()),
+        pa.struct([
+            pa.field('a', 'int8'),
+            pa.field('b', 'string')
+        ]),
+        pa.time32('s'),
+        pa.time64('us'),
+        pa.date32(),
+        pa.date64(),
```
Comment (BryanCutler) [inline python/pyarrow/tests/test_schema.py]: timestamp types should be fine as well right?

**Review by BryanCutler — APPROVED** [review]
+1 thanks!

File: python/pyarrow/tests/test_schema.py (hunk 1 above)
Comment (wesm) [reply python/pyarrow/tests/test_schema.py]: thought I had one, let me add

**Pull request closed without merging**